packages/test-sync/src/androidAndroidTest/kotlin/io/realm/kotlin/test/mongodb/shared/serializer/BsonEncoderTests.kt

```diff
@@ -0,0 +1,321 @@
+            invalidBsonValue = BsonString("")
+        ),
+        WrongTypeAsserter(
+            deserializedType = Short::class,
+            requiredBsonType = BsonNumber::class,
+            invalidBsonValue = BsonString("")
+        ),
+        WrongTypeAsserter(
+            deserializedType = Int::class,
+            requiredBsonType = BsonNumber::class,
+            invalidBsonValue = BsonString("")
+        ),
+        WrongTypeAsserter(
+            deserializedType = Long::class,
+            requiredBsonType = BsonNumber::class,
+            invalidBsonValue = BsonString("")
+        ),
+        WrongTypeAsserter(
+            deserializedType = Float::class,
+            requiredBsonType = BsonNumber::class,
+            invalidBsonValue = BsonString("")
+        ),
+        WrongTypeAsserter(
+            deserializedType = Double::class,
+            requiredBsonType = BsonNumber::class,
+            invalidBsonValue = BsonString("")
+        ),
+        WrongTypeAsserter(
+            deserializedType = Boolean::class,
+            requiredBsonType = BsonBoolean::class,
+            invalidBsonValue = BsonString("")
+        ),
+        WrongTypeAsserter(
+            deserializedType = String::class,
+            requiredBsonType = BsonString::class,
+            invalidBsonValue = BsonInt32(0)
+        ),
+        WrongTypeAsserter(
+            deserializedType = Char::class,
+            requiredBsonType = BsonString::class,
+            invalidBsonValue = BsonInt32(0)
+        ),
+        WrongTypeAsserter(
+            deserializedType = ByteArray::class,
+            requiredBsonType = BsonBinary::class,
+            invalidBsonValue = BsonInt32(0)
+        ),
+    )
+
+    private val realmAsserters = listOf(
+        WrongTypeAsserter(
+            deserializedType = RealmUUID::class,
+            deserializationStrategy = serializer(RealmUUID::class),
+            requiredBsonType = BsonBinary::class,
+            invalidBsonValue = BsonString("")
+        ),
+        WrongTypeAsserter(
+            deserializedType = MutableRealmInt::class,
+            deserializationStrategy = serializer(MutableRealmInt::class),
+            requiredBsonType = BsonInt64::class,
+            invalidBsonValue = BsonString("")
+        ),
+        WrongTypeAsserter(
+            deserializedType = ObjectId::class,
+            deserializationStrategy = serializer(ObjectId::class),
+            requiredBsonType = BsonObjectId::class,
+            invalidBsonValue = BsonString("")
+        ),
+        WrongTypeAsserter(
+            deserializedType = RealmInstant::class,
+            deserializationStrategy = serializer(RealmInstant::class),
+            requiredBsonType = BsonDateTime::class,
+            invalidBsonValue = BsonString("")
+        ),
+    )
+
+    @Test
+    fun decodeFromBsonElement_throwsWrongType() {
+        (primitiveAsserters + realmAsserters).forEach {
+            it.assert()
```

Comment:
I think we should have the `assertFailsWithMessage` here and not burried inside the `WrongTypeAsserter`